analytics/consent_test.go

```diff
@@ -0,0 +1,148 @@
+package analytics
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/stretchr/testify/assert"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"strings"
+	"testing"
+)
+
+func newConsentTempfile(t *testing.T) string {
+	pattern := "section-cli-analytics-consent-" + strings.ReplaceAll(t.Name(), "/", "_")
+	file, err := ioutil.TempFile("", pattern)
+	if err != nil {
+		t.FailNow()
+	}
+	return file.Name()
+}
+
+func TestConsentDetectsIfConsentNotRecorded(t *testing.T) {
+	assert := assert.New(t)
+
+	// Setup
+	consentPath = newConsentTempfile(t)
+
+	// Test
+	assert.False(IsConsentRecorded())
+}
+
+func TestConsentPromptsForConsentIfConsentNotRecorded(t *testing.T) {
+	assert := assert.New(t)
+
+	// Setup
+	consentPath = newConsentTempfile(t)
+
+	var outbuf bytes.Buffer
+	out = &outbuf
+	var inbuf bytes.Buffer
+	inbuf.Write([]byte("\n"))
+	in = &inbuf
+
+	// Invoke
+	ReadConsent()
+
+	// Test
+	assert.Contains(outbuf.String(), "[y/N]")
+}
+
+func TestConsentPromptRecordsConsent(t *testing.T) {
+	assert := assert.New(t)
+
+	// Setup
+	consentPath = newConsentTempfile(t)
+
+	var outbuf bytes.Buffer
+	out = &outbuf
+
+	var inbuf bytes.Buffer
+	inbuf.Write([]byte("y\n"))
+	in = &inbuf
+
+	// Invoke
+	ReadConsent()
+
+	// Test
+	consentFile, err := os.Open(consentPath)
+	assert.NoError(err)
+	contents, err := ioutil.ReadAll(consentFile)
+	assert.NoError(err)
+	var consent cliTrackingConsent
+	err = json.Unmarshal(contents, &consent)
+	assert.NoError(err)
+	assert.True(consent.ConsentGiven)
+}
+
+func TestConsentPromptDefaultsToFalse(t *testing.T) {
+	assert := assert.New(t)
+
+	testCases := []string{
+		"\n",
+		"n\n",
+		"N\n",
+		"OESNAOSENAO\n",
+		"😈\n",
+	}
+	for _, tc := range testCases {
```

Comment:
These tests wouldn't work in parallel because you use global variables. This is not immediately obvious here. I can come in one year and add `t.Parallel()` and break this. Sometimes. So better avoid global vars.

Reply:
I understand. 

In this particular case, given this is a simple CLI tool – not a highly concurrent server – I think it's a reasonable maintainability tradeoff.